elastic/kibana
# [Uptime] implement monitor management telemetry

## Summary

Relates to https://github.com/elastic/kibana/issues/124361

Added telemetry for monitor creation, updates, and deletions.

The schema is as follows

```
  updatedAt?: string; // the date the monitor was created, or updated
  lastUpdatedAt?: string; // the date the monitor was last updated. Omitted when the monitor is first created
  durationSinceLastUpdated?: number; // the time in milliseconds between the last updated and the most recent update. Omitted when the monitor is created.
  deletedAt?: string; // the date the monitor was deleted. Ommitted unless the monitor is deleted
  type: string; // monitor type
  stackVersion: string;
  monitorNameLength: number;
  monitorInterval: number;
  locations: string[];
  countOfLocations: number;
  scriptType?: 'inline' | 'recorder' | 'zip';
  revision: number;
  error?: Array<{ locationId: string; error: Error }>;
  configId: string; // a hash of the monitor id
```

This payload will be transformed into ES documents via a [telemetry indexer and jenkins job](https://github.com/elastic/telemetry/blob/main/kpi-engine/README.md), coming in a separate PR. The schema for those documents will match what is outlined in https://github.com/elastic/kibana/issues/124361

**To test**
--
1. Connect your local version of Kibana to the synthetics service. I like to do so by connecting my Kibana instance to a remote cluster using `oblt-cli`, and then using that remote cluster as the value for `xpack.uptime.service.hosts`.
2. Create a browser monitor. Note the number of steps. Create a second browser monitor with a different number of steps
3. Wait for data to be indexed to Elasticsearch
4. Navigate to Stack Management -> Advanced Settings. Scroll down to Usage data. Click the `cluster data` example.
5. Observe the synthetics data points in the flyout
<img width="828" alt="Screen Shot 2022-02-08 at 5 43 47 PM" src="https://user-images.githubusercontent.com/11356435/153108779-974431d5-5704-43bf-8893-8802871381d5.png">

6. Navigate back to Monitor Managment. Play around with adding, editing, or deleting monitors however you like.
7. Once finished, run both these jobs: [Monitor current channel job](https://internal-ci.elastic.co/blue/organizations/jenkins/elastic%2Btelemetry%2Bmain%2Bsynthetics-monitor-current-staging/activity/) and [Monitor update channel job](https://internal-ci.elastic.co/blue/organizations/jenkins/elastic%2Btelemetry%2Bmain%2Bsynthetics-monitor-update-staging/activity/) This will ensure the data is ingested into the telemetry cluster
8. Visit the [Telemetry dashboard](https://telemetry-v2-staging.elastic.dev/s/synthetics/app/dashboards#/view/6d6c8800-85ca-11ec-9cb9-458dd8a5030c?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-15m,to:now))) to see your data in dashboards
